python/raf/_tvm_op/loss.py

```diff
@@ -83,13 +83,16 @@ def smooth_l1_loss_dtrue_compute(attr, inputs, output_type):  # pylint: disable=
 def nll_loss_compute(attrs, inputs, output_type):  # pylint: disable=unused-argument
     true, pred = inputs
     n, c = pred.shape
+    dtype = pred.dtype
+    if dtype == "float16":
+        pred = pred.astype("float32")
```

Comment:
1. Out of curiosity, what's the difference between `astype` and `cast` in TE compute?
2. If we now allow `nll_loss` to accept float16 inputs and implicitly turn them to float32 for accumulation, can we change the AMP rule accordingly (which forces `nll_loss` inputs to be float32 now)?

Reply:
1. they are the same
2. there are 3 main uses of `nll_loss`, 1) from_pytorch + nll_loss, with amp; 2) from_pytorch + nll_loss, pure fp16 model; 3) ratex lower. Change amp rule only covers 1).

Reply:
For 2, I know that it only covers the first case you mentioned and that's exactly what I meant, although I guess it might not be beneficial because the softmax before it still runs with float32. Anyways, if there's no side effect then we could change that for better consistency.

Reply:
`nll_loss` is already in "never cast"